add: wip operators

nizolnier opened the pull request:
- Fixed formatting
- Added missing bitwise operators
- Added practice problems + answers
- Could use help for bitwise operators' practice

nizolnier committed:
add: wip operators

jerrettl commented on src/intro-to-c/operators/operators-intro.md:
```diff
@@ -0,0 +1,288 @@
+
+## Arithmetic Operators
+An arithmetic operator performs mathematical operations, the ones we learn in elementary school. 
+- **`+`** : addition
+
+- **`-`** : subtraction
+
+- **`*`** : multiplication
+
+- **`/`** : division
+
+- **`%`** : modulus, remainder after division
+
+- **`++`** : increment, increases the value by one
+
+- **`--`** : decrement, decreases the value by one
```
I would remove these spaces so this bullet list matches the other bullet lists

```suggestion
- **`+`** : addition
- **`-`** : subtraction
- **`*`** : multiplication
- **`/`** : division
- **`%`** : modulus, remainder after division
- **`++`** : increment, increases the value by one
- **`--`** : decrement, decreases the value by one
```

jerrettl commented on src/intro-to-c/operators/operators-intro.md:
```diff
@@ -0,0 +1,288 @@
+Bitwise operators perform bit-level operations. This is not used in intro to C, but it will be a good reference for the future.
+- **`&`** : Bitwise AND, copies a bit to the result if it exists in both operands
+- **`|`** : Bitwise OR, copies a bit if it exists in either operand.
+- **`^`** : Bitwise XOR, copies the bit if it is set in one operand but not both.
```
If you want to add it, "XOR" is short for "e**x**clusive **or**"

jerrettl commented on src/intro-to-c/operators/operators-intro.md:
```diff
@@ -0,0 +1,288 @@
+- **`|`** : Bitwise OR, copies a bit if it exists in either operand.
+- **`^`** : Bitwise XOR, copies the bit if it is set in one operand but not both.
+
+So to understand this example, I recommend you have a binary converter opened or if you have taken CDA with Almaki, you will be pretty familiar with binary convertions. 
```
I would remove/edit this line. It refers to a professor that not everyone has taken, especially if these materials become pushed to an audience outside of UCF, though the mild jab at that class is pretty good. If you want to keep the reference to using a binary converter, adding a link to one would be really nice. I found a good one here: https://bitwisecmd.com/#help

jerrettl commented on src/intro-to-c/operators/operators-intro.md:
````diff
@@ -0,0 +1,288 @@
+```
+The output is
+```
+(a == b) && (c > b) is 1 // both expressions are true, so the compound expression is true
+(a == b) && (c < b) is 0 // the first expression is true but the second is not, so the compound expression is false
+(a == b) || (c < b) is 1 // the first expression is true, so the compound expression is true
+(a != b) || (c < b) is 0 // both expressions are false, so the compound expression is false
+!(a != b) is 1 // a != b will return 0 but since we have the !, then !0 is 1
+!(a == b) is 0 // a == b will return 1 but since we have the !, then !1 = 0
````
The comments are really good but I think it makes this look cramped, maybe align the comments so they're a little easier to read

```suggestion
(a == b) && (c > b) is 1     // both expressions are true, so the compound expression is true
(a == b) && (c < b) is 0     // the first expression is true but the second is not, so the compound expression is false
(a == b) || (c < b) is 1     // the first expression is true, so the compound expression is true
(a != b) || (c < b) is 0     // both expressions are false, so the compound expression is false
!(a != b) is 1               // a != b will return 0 but since we have the !, then !0 is 1
!(a == b) is 0               // a == b will return 1 but since we have the !, then !1 = 0
```

jerrettl reviewed (commented):
My thoughts:

- Overall, the content is really comprehensive and covers everything I would imagine for this topic.
- The bitwise potion is great and pretty much exactly what I would've put down. Maybe you could add a breakdown of the first example bit-by-bit just to show how the answer is gotten, but it's explained in the text before it, so it's not entirely necessary. You may also just want to add a title before each example to show what you're about to look at, since it's not entirely clear until you see `a|b` for the first time, for example.
- I agree with moving the ternary operator. It's different from the others, and it makes more sense to showcase it with comparison to an if-else statement.
- In terms of formatting, there are leading whitespaces at the end of a lot of print statements. I would remove them, but it doesn't ultimately affect anything, just my personal preference.
- Please add practice problems if possible. It doesn't have to be a lot, let me know if you need any help with that.
- Something I've noticed with this PR and the others is you never accredit yourself. I'm not sure if that's by choice, and that's okay if it is, but I would be proud of this and put your name on it. 
- See the individual code comments for other specific things.

jerrettl commented:
I just remembered there are three other binary operators, `~` (bitwise NOT), `<<` (bitwise shift left), and `>>` (bitwise shift right). Let me know if you need any help with those.

nizolnier committed:
fix: formatting in operators

nizolnier committed:
add: practice problems and answers

jerrettl commented on src/intro-to-c/operators/operators-practice-answers.md:
```diff
@@ -0,0 +1,171 @@
+
+A fun thing to do in these exercises is to create a separate function and call it in main. That way, you can have your own library of functions and reuse it if you need.
+
+1. Write a function that takes 5 numbers and returns the sum and average (in two decimal places) of them.
```
```suggestion
1. Write a function that takes 5 integers and returns their sum, then write a function that takes 5 integers and returns their average as a float, using the sum function.
```

jerrettl commented on src/intro-to-c/operators/operators-practice-answers.md:
````diff
@@ -0,0 +1,171 @@
+}
+```
+
+2. Write a function that receives a number and returns whether it is even or odd.
````
```suggestion
2. Write a function that takes an integer and prints text stating whether it is even or odd.
```

nizolnier replied:
i feel like this is too specific and doesn't allow for creativity. people can have different solutions to the question and i do not want to limit them by telling them exactly what to do. when i came up with the solution, i am not trying to say that mine is the most ideal one or the exact solution, but it is one way of doing it and that can help someone who was stuck or couldn't think of a way to solve it

jerrettl replied:
Thanks for clarifying here. In that case, I take back my suggestions on re-wording the questions. Instead, I would just suggest putting on the solutions page that these are example solutions and not the "de facto" solutions, so it's clear that it's okay if you got something different.

jerrettl commented on src/intro-to-c/operators/operators-practice-answers.md:
````diff
@@ -0,0 +1,171 @@
+}
+```
+
+3. Write a function that receives a number and returns if the number is divisible by 3.
````
```suggestion
3. Write a function that takes an integer and prints text stating if it is divisible by 3.
```

jerrettl commented on src/intro-to-c/operators/operators-practice-answers.md:
````diff
@@ -0,0 +1,171 @@
+}
+```
+
+4. Write a function that takes two numbers and returns which one is higher.
````
```suggestion
4. Write a function that takes two integers and returns the greater of the two. (If they are equal, you may return either.)
```

jerrettl commented on src/intro-to-c/operators/operators-practice-answers.md:
````diff
@@ -0,0 +1,171 @@
+
+```
+
+5. Write a function that takes three numbers, the sides of a triangle and returns whether the triangle formed by the given sides is equilateral, isosceles, or scalene.
````
```suggestion
5. Write a function that takes three integers, representing the lengths of the sides of a triangle, and prints text stating what type of triangle (equilateral, isosceles, or scalene) is formed by the given lengths. *(An equilateral triangle is one where all sides are equal length. An isosceles triangle is one where exactly two sides are equal length. A scalene triangle is one where all sides are different lengths.)*
```

nizolnier replied:
i think if you're at this point of a cs degree, you should have the basic knowledge to know what are the types of a triangle

jerrettl replied:
Honestly, I wrote this because even I forgot what isosceles meant, so I'm sure others have forgotten too. Obviously these are not hard concepts, but a quick reminder would be nice for those that have forgotten.

nizolnier replied:
google is free and available for everyone

jerrettl replied:
Fair enough

jerrettl commented on src/intro-to-c/operators/operators-practice-answers.md:
```diff
@@ -0,0 +1,171 @@
+        printf("Isosceles Triangle\n");
+    } // a scalene triangle has all different sides
+    else {
+        printf("Scalene triangle\n");
```
The other print statements capitalize the T
```suggestion
        printf("Scalene Triangle\n");
```

jerrettl commented on src/intro-to-c/operators/operators-practice.md:
```diff
@@ -1 +1,13 @@
+
+4. Write a function that takes two numbers and returns which one is higher.
+
+5. Write a function that takes three numbers, the sides of a triangle and returns whether the triangle formed by the given sides is equilateral, isosceles, or scalene.
```
```suggestion
1. Write a function that takes 5 integers and returns their sum, then write a function that takes 5 integers and returns their average as a float, using the sum function.

2. Write a function that takes an integer and prints text stating whether it is even or odd.

3. Write a function that takes an integer and prints text stating if it is divisible by 3.

4. Write a function that takes two integers and returns the greater of the two. (If they are equal, you may return either.)

5. Write a function that takes three integers, representing the lengths of the sides of a triangle, and prints text stating what type of triangle (equilateral, isosceles, or scalene) is formed by the given lengths. *(An equilateral triangle is one where all sides are equal length. An isosceles triangle is one where exactly two sides are equal length. A scalene triangle is one where all sides are different lengths.)*
```

jerrettl reviewed (commented):
- The new additions to the operators introduction look good.

- I left some suggestions on rewording the questions to better reflect the solutions.

- Something I notice is that the questions ask for a function, however there is also a main function added in the solution. Not sure if the intent is to also write that, or if you would already assume that a pre-filled main function exists, but I would clarify if that is the case either way. If it's the first case, just a note in the solutions file; if it's the second case, maybe add the pre-filled main function in the questions file so you have an idea on how the function you write will be used.

- Since questions 2-5 use conditionals, you may want to add a short label that these are more "challenging." (something super basic like [this](https://github.com/Wiki-Knights/CS-Materials/blob/f0239be2e0fed796726617852509c84ac24014d9/src/intro-to-c/functions/functions-practice.md?plain=1#L96) or for a little more fancy like [this](https://github.com/Wiki-Knights/CS-Materials/blob/f0239be2e0fed796726617852509c84ac24014d9/src/intro-to-c/conditionals/conditionals-practice.md?plain=1#L146)) I say this because you'll probably learn about operators in class before you learn about conditionals, so while these are good questions, I just want to make sure it's noted for students that might be following along with these materials as they learn in class.

    If you feel like there's not enough non-challenge problems with that change, you could add problems that are like "write this mathematical expression as code" where you the mathematical expression is something like:

    ![image](https://user-images.githubusercontent.com/7614459/166552285-2a87d261-8247-4a7c-86f2-8940a63d3330.png)

- For adding practice on bitwise operators, these are kind of hard to have any examples until you get to a class that covers concepts like bitmasking, but I guess you could have similar problems that are like, "convert the following boolean expression into code," where the expression is like "NOT x AND (a OR b) AND (c AND d)". Not super complicated to solve, but gives practice using them nonetheless.

- Maybe for the shift left and shift right operators you could have a challenge problem or two saying like "multiply (an integer) n by 8 with only bitwise operators" (`n << 3`) or "divide (an integer) n by 2 with only bitwise operators," (`n >> 1`) since shifting left x times will create the same result as multiplying a number by 2<sup>x</sup>, and the opposite for dividing.

Let me know if you need any help making any of these. Now that we're done with classes, I'll have some more free time again to put towards this.

nizolnier committed:
add: alterations

Commit:
Metadata, Whitespace, Lists updates

Added some newlines here and there, removed trailing whitespace here and there. Made numbered lists auto-numbered. Added metadata blocks.

ZThatWhiteGuyZ committed:
Merge branch 'main' into add/arithmetic-operators

Commit:
Update index.md

Table of Contents update for arithmetic operators.

Commit:
Question Updates

Added two new questions. One dealing with bit shifting, and one with writing slightly more elaborate mathematical expressions in code.  Also added myself to the author list.

Commit:
Update operators-practice-answers.md

Correction to last answer.
